lib/chat_models/chat_anthropic.ex

```diff
@@ -0,0 +1,560 @@
+      role: :assistant,
+      content: "",
+      status: stop_reason_to_status(stop_reason)
+    }
+    |> MessageDelta.new()
+    |> to_response()
+  end
+
+  # These streaming events don't contain any useful data to return to callers. Ignore.
+  def do_process_response(%{"type" => type})
+      when type in ["message_start", "ping", "content_block_stop", "message_stop"] do
+    :ignore
+  end
+
+  def do_process_response(%{"error" => %{"message" => reason}}) do
+    Logger.error("Received error from API: #{inspect(reason)}")
+    {:error, reason}
+  end
+
+  def do_process_response({:error, %Jason.DecodeError{} = response}) do
+    error_message = "Received invalid JSON: #{inspect(response)}"
+    Logger.error(error_message)
+    {:error, error_message}
+  end
+
+  def do_process_response(other) do
+    Logger.error("Trying to process an unexpected response. #{inspect(other)}")
+    {:error, "Unexpected response"}
+  end
+
+  defp to_response({:ok, message}), do: message
+  defp to_response({:error, changeset}), do: {:error, Utils.changeset_error_to_string(changeset)}
+
+  defp stop_reason_to_status("end_turn"), do: :complete
+  defp stop_reason_to_status("max_tokens"), do: :length
+  defp stop_reason_to_status("stop_sequence"), do: :complete
+
+  defp stop_reason_to_status(other) do
+    Logger.warning("Unsupported stop_reason. Reason: #{inspect(other)}")
+    nil
+  end
+
+  defmodule StreamingChunkDecoder do
+    @moduledoc false
+
+    # Data comes back from Anthropic in event/data pairs. See the following example:
+    #
+    #     event: message_start
+    #     data: {"type": "message_start", "message": {"id": "msg_1nZdL29xx5MUA1yADyHTEsnR8uuvGzszyY", "type": "message", "role": "assistant", "content": [], "model": "claude-3-opus-20240229", "stop_reason": null, "stop_sequence": null, "usage": {"input_tokens": 25, "output_tokens": 1}}}
+    #
+    #     event: content_block_start
+    #     data: {"type": "content_block_start", "index": 0, "content_block": {"type": "text", "text": ""}}
+    #
+    #     event: ping
+    #     data: {"type": "ping"}
+    #
+    #     event: content_block_delta
+    #     data: {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "Hello"}}
+    #
+    #     event: content_block_delta
+    #     data: {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "!"}}
+    #
+    #     event: content_block_stop
+    #     data: {"type": "content_block_stop", "index": 0}
+    #
+    #     event: message_delta
+    #     data: {"type": "message_delta", "delta": {"stop_reason": "end_turn", "stop_sequence":null, "usage":{"output_tokens": 15}}}
+    #
+    #     event: message_stop
+    #     data: {"type": "message_stop"}
+    #
+    #
+    # See Anthropic documentation: https://docs.anthropic.com/claude/reference/messages-streaming
+
+    # This captures `event` and `data`
+    @event_data_re ~r/^\s*event:\s*(?<event>\S*)\s*data:\s*(?<data>[^\n]+)\s*$/
```

Comment:
I'm wondering if we could do binary pattern matches on the event type and data instead of RegEx?

We'd need a catch-all to future proof from news events being added in the future.

As an example, it might look like this:

```elixir
def decode("event: message_start\ndata: " <> data) do
  # process data
end

def decode("event: ping\n" <> _rest), do: :ignore

# ... more events we handle or ignore ...

def decode("event: " <> rest) do
  Logger.error("Unsupported event received")
  :ignore
end
```

Something like this is very fast to execute but, more importantly, more clear and explicit.

You know more about this protocol. Is this a possible solution?

Reply:
So I had done something like this in my own code base with OpenAI originally and I will say that there have been a few times where we noticed an extra or a missing whitespace/newline which broke our parser. That seems to break their contract, so to be fair I assume that is a bug on their end. While that was openai, I wouldn't be surprised to see the same thing happen with other APIs. The goal of the regex was to be a bit less sensitive while parsing.

I agree the pattern matching solution reads much nicer though so I can switch to that